pkg/controller/client_builder_dynamic.go

```diff
@@ -0,0 +1,207 @@
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/transport"
+	"k8s.io/klog"
+	utilpointer "k8s.io/utils/pointer"
+)
+
+// defaultExpirationSeconds defines the duration of a TokenRequest in seconds.
+var defaultExpirationSeconds = int64(3600)
+
+type DynamicControllerClientBuilder struct {
+	// ClientConfig is a skeleton config to clone and use as the basis for each controller client
+	ClientConfig *restclient.Config
+
+	// CoreClient is used to provision service accounts if needed and watch for their associated tokens
+	// to construct a controller client
+	CoreClient v1core.CoreV1Interface
+
+	// AuthenticationClient is used to check API tokens to make sure they are valid before
+	// building a controller client from them
+	AuthenticationClient v1authentication.AuthenticationV1Interface
+
+	// Namespace is the namespace used to host the service accounts that will back the
+	// controllers.  It must be highly privileged namespace which normal users cannot inspect.
+	Namespace string
+
+	// roundTripperFuncMap is a cache stores the corresponding roundtripper func for each
+	// service account
+	roundTripperFuncMap map[string]func(http.RoundTripper) http.RoundTripper
+
+	expirationSeconds int64
+
+	mutex sync.Mutex
+
+	clock clock.Clock
+}
+
+func NewDynamicClientBuilder(clientConfig *restclient.Config, coreClient v1core.CoreV1Interface, authClient v1authentication.AuthenticationV1Interface, ns string) ControllerClientBuilder {
+	builder := &DynamicControllerClientBuilder{
+		ClientConfig:         clientConfig,
+		CoreClient:           coreClient,
+		AuthenticationClient: authClient,
+		Namespace:            ns,
+		roundTripperFuncMap:  map[string]func(http.RoundTripper) http.RoundTripper{},
+		expirationSeconds:    defaultExpirationSeconds,
+		clock:                clock.RealClock{},
+	}
+	return builder
+}
+
+// this function only for test purpose, don't call it
+func NewTestDynamicClientBuilder(clientConfig *restclient.Config, coreClient v1core.CoreV1Interface, authClient v1authentication.AuthenticationV1Interface, ns string, expirationSeconds int64) ControllerClientBuilder {
+	builder := &DynamicControllerClientBuilder{
+		ClientConfig:         clientConfig,
+		CoreClient:           coreClient,
+		AuthenticationClient: authClient,
+		Namespace:            ns,
+		roundTripperFuncMap:  map[string]func(http.RoundTripper) http.RoundTripper{},
+		expirationSeconds:    expirationSeconds,
+		clock:                clock.RealClock{},
+	}
+	return builder
+}
+
+func (t *DynamicControllerClientBuilder) Config(saName string) (*restclient.Config, error) {
+	_, err := getOrCreateServiceAccount(t.CoreClient, t.Namespace, saName)
+	if err != nil {
+		return nil, err
+	}
+
+	configCopy := constructClient(t.Namespace, saName, t.ClientConfig)
+
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	rt, ok := t.roundTripperFuncMap[saName]
+	if ok {
+		configCopy.WrapTransport = rt
+	} else {
+		dur := time.Duration(t.expirationSeconds*20/100) * time.Second
```

Comment:
don't use leeway to compute the 20% renew period. the server is free to return tokens valid for shorter periods than we ask for. compute 80% for a particular token when we set Expiry